### PR TITLE
Separate dependencies and Telegram build to jobs

### PR DESCRIPTION
**Pull request by RadRussianRus**
This PR addresses the problem with rebuilding on CI.

## About the problem
When CI action fails for whatever reason, next time it will be fully rebuilt, both dependencies and Telegram. This leads to increased build times due to unnecessary rebuilds of dependencies.

## Proposed solution
My currently proposed solution is to split dependencies and Telegram build to separate jobs. That way even if Telegram build will fail, dependencies will be cached and reused on next run.

#### Why not use `continue-on-error`?
`continue-on-error` will make job marked successfully done even when it fails, so finding errors will be harder.

### CODE REVIEW

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -411,6 +372,21 @@ jobs:
           cd ..
           rm -rf libxkbcommon
 
+      - name: Libwayland.
```
Comment (ilya-fedin) [inline .github/workflows/linux.yml:375]: Why did you move this?

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -42,32 +42,30 @@ on:
...
-    name: Ubuntu 14.04
+  dependencies:
+    name: "Ubuntu 14.04: dependencies"
     runs-on: ubuntu-latest
     container: ubuntu:trusty
 
     strategy:
       matrix:
         defines:
           - ""
```
Comment (ilya-fedin) [inline .github/workflows/linux.yml:68]: You can remove it since it is not used

File: .github/workflows/linux.yml [hunk 3]
```diff
@@ -111,10 +109,10 @@ jobs:
           sudo update-alternatives --config gcc && \
           sudo add-apt-repository --remove ppa:ubuntu-toolchain-r/test -y
 
+      # We don't need recursive clone to build dependencies.
       - name: Clone.
         uses: actions/checkout@v2
         with:
-          submodules: recursive
           path: ${{ env.REPO_NAME }}
```
Comment (ilya-fedin) [inline .github/workflows/linux.yml:116]: And this too

Comment (RadRussianRus) [reply .github/workflows/linux.yml:116]: No, it's needed. Since we're generating cache key from current workflow file, we need at least shallow clone of repo.

File: .github/workflows/linux.yml (hunk 2 above)
Comment (RadRussianRus) [reply .github/workflows/linux.yml]: I thought of it as a placeholder, like on other workflows, but since there are not "Read defines" for them, you're probably right. Removed in 92b261b.

File: .github/workflows/linux.yml (hunk 1 above)
Comment (RadRussianRus) [reply .github/workflows/linux.yml]: Oops, missed that part. Fixed in 8d35320.

File: .github/workflows/win.yml [hunk 4]
```diff
@@ -399,6 +374,154 @@ jobs:
...
+          echo Find any version of Python 2.
+          for /D %%a in (C:\hostedtoolcache\windows\Python\2.*) do (
+            SET PY2=%%a\x64
+          )
+          if [%PY2%] == [] (
+            echo Python 2 is not found.
+            exit 1
+          )
+          echo Found %PY2%.
+          echo ::set-env name=PY2::%PY2%
```
Comment (23rd) [inline .github/workflows/win.yml:453]:
```suggestion
```

**Comment by 23rd** [comment]
I don't like to see so much repeating code.

I have opened a new pull request to your branch, but it's not enough.
For example, each job installs the same packages now. This should be reduced. Each job should only install a necessary set of packages.

**Comment by RadRussianRus** [comment]
> I don't like to see so much repeating code.
> 
> I have opened a new pull request to your branch, but it's not enough.
> For example, each job installs the same packages now. This should be reduced. Each job should only install a necessary set of packages.

Agree, that should be reduced as much as possible. I'll test later on how this could be reduced.

**Review by VChernomYA — APPROVED** [review]

**Review by VChernomYA — APPROVED** [review]

File: .github/workflows/linux.yml (hunk 1 above)
Comment (VChernomYA) [reply .github/workflows/linux.yml]: https://github.com/telegramdesktop/tdesktop/commit/8d353202a4970e5c12e0050531662904b37f1bb3

**Comment by john-preston** [comment]
@RadRussianRus Is this still relevant? Or could it be closed?

**Comment by RadRussianRus** [comment]
I guess it should be closed since CI have changed a lot. I'll open a new repo if needed.
